utilities/qneg.F90

```diff
@@ -0,0 +1,509 @@
+module qneg
+
+  use shr_kind_mod,        only: CS => SHR_KIND_CS
+  use ccpp_kinds,          only: kind_phys
+  use perf_mod,            only: t_startf, t_stopf
+  use cam_logfile,         only: iulog
+  use cam_abortutils,      only: endrun, check_allocate
+  use shr_sys_mod,         only: shr_sys_flush
+!  use cam_history_support, only: max_fieldname_len
+!  use ppgrid,              only: pcols
+
+  implicit none
+  private
+  save
+
+  ! Public interface.
+
+  public :: qneg_init
+  public :: qneg_run !qneg3 in CAM6
+  public :: qneg_timestep_final
+  public :: qneg_final
+!  public :: qneg4
+
+  ! Private module variables
+  logical          :: collect_stats = .false.
+  logical          :: timestep_reset = .false.
+  logical          :: qneg_initialized = .false.
+  integer          :: num_constituents = 0
+
+  real(kind_phys), parameter :: tol = 1.e-12_kind_phys
+  real(kind_phys), parameter :: worst_reset = 1.e35_kind_phys
+
+  ! Diagnostic field names
+!  integer              :: num_diag_fields = 0
+!  character(len=max_fieldname_len) :: diag_names(num_diag_fields)
+!  logical             :: cnst_out_calc = .false.
+!  logical             :: cnst_outfld(num_diag_fields) = .false.
+
+  ! Summary buffers
+  integer, parameter    :: num3_bins = 24
+!  integer, parameter    :: num4_bins = 4
+  character(len=CS)     :: qneg3_warn_labels(num3_bins) = ''
+!  character(len=CS)     :: qneg4_warn_labels(num4_bins) = ''
+  integer, allocatable  :: qneg3_warn_num(:,:)
+!  integer               :: qneg4_warn_num(num4_bins)          = 0
+  real(kind_phys), allocatable :: qneg3_warn_worst(:,:)
+!  real(kind_phys)              :: qneg4_warn_worst(num4_bins)        = worst_reset
+
+  private             :: qneg_print_summary
+!  private             :: calc_cnst_out
+  private             :: find_index3
+!  private             :: find_index4
+  interface reset_stats
+     module procedure reset_stats_scalar
+     module procedure reset_stats_array
+  end interface reset_stats
+
+contains
+
+!> \section arg_table_qneg_init Argument Table
+!! \htmlinclude qneg_init.html
+  subroutine qneg_init(print_qneg_warn, number_of_constituents, qmin)
```

Comment:
Should have `errcode` and `errmsg` (initialize them to zero and empty string respectively).

Reply:
good call - added in both routines